src/main/scala/chisel3/util/Inline.scala

```diff
@@ -0,0 +1,96 @@
+// See LICENSE for license details.
+
+package chisel3.util
+
+import chisel3._
+import chisel3.internal.InstanceId
+import chisel3.experimental.{BaseModule, ChiselAnnotation, RunFirrtlTransform}
+import firrtl.passes.{InlineAnnotation, InlineInstances}
+import firrtl.transforms.NoDedupAnnotation
+import firrtl.annotations.{CircuitName, ModuleName, ComponentName}
+
+/** Inlines an instance of a [[chisel3.experimental.BaseModule]]
+  *
+  * @example {{{
+  *  /* Provides some dummy module internals */
+  *  trait Internals { this: Module =>
+  *    val io = IO(new Bundle{ val a = Input(Bool()) })
+  *  }
+  *  /* InlineInstance is mixed directly into Foo's definition. Every instance
+  *   * of this will be inlined. */
+  *  class Foo extends Module with Internals with InlineInstance
+  *  /* Bar will, by default, not be inlined */
+  *  class Bar extends Module with Internals
+  *  class Top extends Module {
+  *    val io = IO(new Bundle{ val a = Input(Bool()) })
+  *    val x = Module(new Foo)                     // x will be inlined
+  *    val y = Module(new Bar with InlineInstance) // y will also be inlined
+  *    val z = Module(new Bar)                     // z will not be inlined
+  *    Seq(x, y, z).map(_.io.a := io.a)
+  *  }
+  * }}}
+  */
+trait InlineInstance { self: BaseModule =>
+  chisel3.experimental.annotate(
+    new ChiselAnnotation with RunFirrtlTransform {
+      def toFirrtl = InlineAnnotation(self.toNamed)
+      def transformClass = classOf[InlineInstances] } )
+  chisel3.experimental.annotate(
+    new ChiselAnnotation {
+      def toFirrtl = NoDedupAnnotation(self.toNamed) } )
+}
+
+/** Provide methods for controlling inlining from within a [[chisel3.experimental.BaseModule]]
+  */
+trait InlineHelpers { self: BaseModule =>
+  /* If used anywhere, this will result in every module in the design being
+   * inlined. The resulting circuit will contain only one module. */
+
+  /** This, if used anywhere, will result in the entire circuit being
+    * flattened to a single module.
+    *
+    * @example {{{
+    * /* Providing some module internals */
+    * trait Internals { this: Module =>
+    *   val io = IO(new Bundle{ val a = Input(Bool()) })
+    * }
+    * class Bar extends Module with Internals with InlineHelpers
+    * class Top extends Module with InlineHelpers {
+    *   val io = IO(new Bundle{ val a = Input(Bool()) })
+    *   val Seq(x, y, z) = Seq.fill(3)(Module(new Bar))
+    *   Seq(x, y, z).map(_.io.a := io.a)
+    *   /* All modules in the entire circuit will be inlined. Here, that is x, y,
+    *    * and z. */
+    *   inlineCircuit()
+    * }
+    * }}}
+    */
+  def inlineCircuit(): Unit = chisel3.experimental.annotate(
```

Comment:
This one feels kind of dangerous. I dunno how I feel about an arbitrary point in the Module hierarchy being able to instigate inlining *all* modules. I would maybe make `InlineCircuit` as `trait` and have it error unless it's mixed in to the top module. Alternatively, I think `Flatten` in FIRRTL inlines everything below a given module? Maybe make a `FlattenHierarchy` trait that can be mixed in to flatten everything below and thus Inlining the whole circuit can be accomplished by mixing that in at the top?

Reply:
I personally don't like it either... This PR only includes that because the FIRRTL inlining annotations support a `CircuitName` that will flatten the whole thing. I was struggling with a good API for that at the Chisel level (and will usually aim to expose whatever FIRRTL API exists at the Chisel level). I like the approach of requiring that this be mixed into the top module if that is the desired effect. Or: that's much better than "spooky inlining at a distance..."

This does bring up a point, which did not occur to me, that a Chisel flattening API should be a part of the inlining API (since they're both frontends for `InlineInstances`). Are you in agreement that that would round out this PR?

Reply:
Let me check that I'm understanding correctly but I think I agree.

By Chisel flattening API do you mean some API in Chisel (eg. `trait FlattenHierarchy`) to flatten everything below the given module? And in adding that, we would remove `def inlineCircuit()`?

Reply:
(Sorry, dropped off on this...)

Yes, I mean an API for flattening everything below a given module (an annotator that emits a `firrtl.transforms.FlattenAnnotation` which will run `firrtl.transforms.Flatten` should be all that's needed). I'm fine with removing the `inlineCircuit` API as it seems at worst dangerous (somebody uses this in their library) and at best confusing (SO posts asking, "Why is my circuit inlined?"). The initial motivation for `inlineCircuit` was only to fully expose the entire FIRRTL API of `firrtl.passes.Inline`, but it's far better to expose a usable sane Chisel API.